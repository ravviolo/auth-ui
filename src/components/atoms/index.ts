export { Button } from './Button/Button';
export { Checkbox } from './Checkbox/Checkbox';
export { Divider } from './Divider/Divider';
export { Heading } from './Heading/Heading';
export { IconButton } from './IconButton/IconButton';
export { InputField } from './InputField/InputField';
export { InputLabel } from './InputLabel/InputLabel';
export { TextButton } from './TextButton/TextButton';
