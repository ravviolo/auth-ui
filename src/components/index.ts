export * from './templates/LoginTemplate/LoginTemplate';
export * from './templates/SignUpTemplate/SignUpTemplate';
