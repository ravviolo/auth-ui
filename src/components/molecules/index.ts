export { AuthFooter } from './AuthFooter/AuthFooter';
export { AuthHeader } from './AuthHeader/AuthHeader';
export { AuthSocials } from './AuthSocials/AuthSocials';
export { EmailInput } from './EmailInput/EmailInput';
export { PasswordInput } from './PasswordInput/PasswordInput';
