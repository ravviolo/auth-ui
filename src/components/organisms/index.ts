export * from './LoginForm/LoginForm';
export * from './SignUpForm/SignUpForm';
