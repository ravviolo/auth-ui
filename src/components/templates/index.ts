export * from './LoginTemplate/LoginTemplate';
export * from './SignUpTemplate/SignUpTemplate';
