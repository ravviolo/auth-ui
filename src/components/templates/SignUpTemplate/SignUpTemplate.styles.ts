import styled from 'styled-components';

import { Flex } from 'components/atoms';

export const Container = styled.main`
  width: 25vw;
  height: 70vh;
  padding: 3em;
  background-color: #fff;
  border-radius: 10px;
  box-shadow: 2px 2px 8px 1px rgba(0, 0, 0, 0.3);
`;

export const TemplateFlex = styled(Flex)`
  gap: 2em;
`;
