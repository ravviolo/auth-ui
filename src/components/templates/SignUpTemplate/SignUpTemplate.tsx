import { AuthHeader } from 'components/molecules';
import { SignUpForm } from 'components/organisms';

interface Props {
  onSubmit: (email: string, password: string) => void;
  onClick: () => void;
  onClickGoogle: () => void;
  onClickFacebook: () => void;
  onClickLinkedIn: () => void;
}

export const SignUpTemplate = ({
  onSubmit,
  onClick,
  onClickFacebook,
  onClickGoogle,
  onClickLinkedIn,
}: Props) => {
  return (
    <div>
      <AuthHeader title="Sign Up" />
      <SignUpForm onSubmit={onSubmit} />
    </div>
  );
};
