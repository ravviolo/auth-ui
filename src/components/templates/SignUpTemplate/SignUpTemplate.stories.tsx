import { ComponentMeta, ComponentStory } from '@storybook/react';

import { SignUpTemplate } from 'components/templates';
import { RouterDecorator, RouterLoggerDecorator } from 'storybook/decorators';

export default {
  title: 'Templates/SignUpTemplate',
  component: SignUpTemplate,
  decorators: [RouterLoggerDecorator, RouterDecorator],
  args: {
    footerLinkHref: '/pathname',
    footerLinkText: 'Link',
    footerText: 'Footer text.',
    headerText: 'Header Title',
  },
} as ComponentMeta<typeof SignUpTemplate>;

export const Default: ComponentStory<typeof SignUpTemplate> = (args) => (
  <SignUpTemplate {...args} />
);
Default.storyName = 'SignUpTemplate';
