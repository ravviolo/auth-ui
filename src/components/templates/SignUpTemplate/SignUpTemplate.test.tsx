import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { create } from 'react-test-renderer';

import { SignUpTemplate } from 'components/templates';

import { Props as SignUpTemplateProps } from './SignUpTemplate';

describe('SignUpTemplate', () => {
  const mockOnClickFacebook = jest.fn<
    SignUpTemplateProps['onClickFacebook'],
    []
  >();
  const mockOnClickGoogle = jest.fn<SignUpTemplateProps['onClickGoogle'], []>();
  const mockOnClickLinkedIn = jest.fn<
    SignUpTemplateProps['onClickLinkedIn'],
    []
  >();
  const mockOnSubmitSignUpForm = jest.fn<
    SignUpTemplateProps['onSubmitSignUpForm'],
    [string, string]
  >();
  const mockOnClickFooterBtn = jest.fn<
    SignUpTemplateProps['onClickFooterBtn'],
    []
  >();

  const testEmail = 'test@test.com';
  const testPassword = 'testpassword';

  const props: SignUpTemplateProps = {
    testId: 'test-signup-template',
    headerText: 'SignUp Header',
    footerText: 'SignUp Footer',
    footerBtnText: 'SignUp Footer Button',
    onClickFacebook: mockOnClickFacebook,
    onClickGoogle: mockOnClickGoogle,
    onClickLinkedIn: mockOnClickLinkedIn,
    onSubmitSignUpForm: mockOnSubmitSignUpForm,
    onClickFooterBtn: mockOnClickFooterBtn,
  };

  it('should render signup header with title', () => {
    render(<SignUpTemplate {...props} />);

    const heading = screen.getByTestId('auth-header-signup-template');

    expect(heading).toHaveTextContent('SignUp Header');
  });

  it('should render signup form', () => {
    render(<SignUpTemplate {...props} />);

    const signUpForm = screen.getByTestId('signup-form');

    expect(signUpForm).toBeInTheDocument();
  });

  it('should submit signup form with non-empty input fields', async () => {
    render(<SignUpTemplate {...props} />);

    const signUpForm = screen.getByTestId('signup-form');

    const emailInputField = within(signUpForm).getByTestId('input-field-email');
    const passwordInputField = within(signUpForm).getByTestId(
      'input-field-password'
    );
    const submitSignUpBtn = within(signUpForm).getByTestId('btn-submit-signup');

    await userEvent.type(emailInputField, testEmail);
    await userEvent.type(passwordInputField, testPassword);
    await userEvent.click(submitSignUpBtn);

    expect(mockOnSubmitSignUpForm).toHaveBeenCalledTimes(1);
    expect(mockOnSubmitSignUpForm).toHaveBeenCalledWith(
      testEmail,
      testPassword
    );
  });

  it("shouldn't submit form with one or more empty input fields", async () => {
    render(<SignUpTemplate {...props} />);

    const signUpForm = screen.getByTestId('signup-form');
    const submitSignUpBtn = within(signUpForm).getByTestId('btn-submit-signup');

    await userEvent.click(submitSignUpBtn);

    expect(mockOnSubmitSignUpForm).not.toHaveBeenCalled();

    const emailInputField = within(signUpForm).getByTestId('input-field-email');
    await userEvent.type(emailInputField, testEmail);
    await userEvent.click(submitSignUpBtn);

    expect(mockOnSubmitSignUpForm).not.toHaveBeenCalled();
  });

  it("should render a section divider with 'OR' badge", () => {
    render(<SignUpTemplate {...props} />);

    const divider = screen.getByTestId('divider-signup');

    expect(divider).toHaveTextContent('OR');
  });

  it('should render authenticate through socials section, should display icons for Facebook, Google, LinkedIn', () => {
    render(<SignUpTemplate {...props} />);

    const authSocials = screen.getByTestId('auth-socials-signup-template');

    const authSocialsIcons = within(authSocials).getAllByTestId(/icon-btn-/);

    expect(authSocialsIcons).toHaveLength(3);
  });

  it('should handle click events on Facebook icon - run handler once', async () => {
    render(<SignUpTemplate {...props} />);

    const authSocials = screen.getByTestId('auth-socials-signup-template');
    const facebookIcon = within(authSocials).getByTestId(/icon-btn-facebook/i);

    await userEvent.click(facebookIcon);

    expect(mockOnClickFacebook).toHaveBeenCalledTimes(1);
  });

  it('should handle click events on Google icon - run handler once', async () => {
    render(<SignUpTemplate {...props} />);

    const authSocials = screen.getByTestId('auth-socials-signup-template');
    const googleIcon = within(authSocials).getByTestId(/icon-btn-google/i);

    await userEvent.click(googleIcon);

    expect(mockOnClickGoogle).toHaveBeenCalledTimes(1);
  });

  it('should handle click events on LinkedIn icon - run handler once', async () => {
    render(<SignUpTemplate {...props} />);

    const authSocials = screen.getByTestId('auth-socials-signup-template');
    const linkedInIcon = within(authSocials).getByTestId(/icon-btn-linkedIn/i);

    await userEvent.click(linkedInIcon);

    expect(mockOnClickLinkedIn).toHaveBeenCalledTimes(1);
  });

  it('should render signup footer with text and button', () => {
    render(<SignUpTemplate {...props} />);

    const footer = screen.getByTestId('auth-footer-signup-template');
    const footerBtn = within(footer).getByTestId('btn-footer');

    expect(footer).toHaveTextContent('SignUp Footer');
    expect(footerBtn).toHaveTextContent('SignUp Footer Button');
  });

  it('should handle click events on footer button - run handler once', async () => {
    render(<SignUpTemplate {...props} />);

    const footer = screen.getByTestId('auth-footer-signup-template');
    const footerBtn = within(footer).getByTestId('btn-footer');

    await userEvent.click(footerBtn);

    expect(mockOnClickFooterBtn).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot', () => {
    const tree = create(<SignUpTemplate {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
