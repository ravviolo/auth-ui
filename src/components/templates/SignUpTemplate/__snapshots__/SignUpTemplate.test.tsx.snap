// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpTemplate should match snapshot 1`] = `
<main
  className="sc-dmRaPn dQxFcV"
  data-testid="test-signup-template"
>
  <div
    className="sc-jSMfEi bQrDqy sc-kgflAQ iSfrTS"
    data-testid="flex-signup-template"
    direction="column"
  >
    <header
      data-testid="auth-header-signup-template"
    >
      <h1
        className="sc-gKXOVf zAjwy"
        data-testid="auth-heading"
      >
        SignUp Header
      </h1>
    </header>
    <form
      data-testid="signup-form"
      onSubmit={[Function]}
    >
      <div
        className="sc-jSMfEi bQrDqy sc-idiyUo wwKtw"
        data-testid="flex-signup-form"
        direction="column"
      >
        <div
          className="sc-jSMfEi bQrDqy sc-breuTD jjaaRC"
          data-testid="email-input-signup-form"
          direction="column"
        >
          <label
            className="sc-papXJ daxRFt"
            data-testid="input-label-email"
            htmlFor="email"
          >
            Email
          </label>
          <input
            className="sc-ftvSup dLJLzu"
            data-testid="input-field-email"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div
          className="sc-jSMfEi bQrDqy sc-ksZaOG cRa-dNN"
          data-testid="password-input-signup-form"
          direction="column"
        >
          <label
            className="sc-papXJ daxRFt"
            data-testid="input-label-password"
            htmlFor="password"
          >
            Password
          </label>
          <input
            className="sc-ftvSup dLJLzu"
            data-testid="input-field-password"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <button
          className="sc-bczRLJ jXOWFG"
          data-testid="btn-submit-signup"
          type="submit"
        >
          Sign Up
        </button>
      </div>
    </form>
    <div
      className="sc-dkzDqf fBSlBa"
      data-testid="divider-signup"
    >
      <div
        className="sc-hKMtZM cJojKy"
      />
      <span
        className="sc-eCYdqJ fGmLST"
      >
        OR
      </span>
    </div>
    <div
      className="sc-jSMfEi iChCMC sc-evZas fBPmfS"
      data-testid="auth-socials-signup-template"
      direction="row"
    >
      <button
        className="sc-iBkjds cxbDaZ"
        color="#DB4437"
        data-testid="icon-btn-Google"
        onClick={[MockFunction]}
        type="button"
      >
        G
      </button>
      <button
        className="sc-iBkjds bqhZzF"
        color="#4267B2"
        data-testid="icon-btn-Facebook"
        onClick={[MockFunction]}
        type="button"
      >
        f
      </button>
      <button
        className="sc-iBkjds hAqsSt"
        color="#0077B5"
        data-testid="icon-btn-LinkedIn"
        onClick={[MockFunction]}
        type="button"
      >
        in
      </button>
    </div>
    <footer
      className="sc-kDDrLX kktpMP"
      data-testid="auth-footer-signup-template"
    >
      <div
        className="sc-jSMfEi iChCMC sc-crXcEl bUmuaF"
        data-testid="flex-footer"
        direction="row"
      >
        <p>
          SignUp Footer
        </p>
        <button
          className="sc-jqUVSM ebpvV sc-iqcoie bPYbbt"
          data-testid="btn-footer"
          onClick={[MockFunction]}
          type="button"
        >
          SignUp Footer Button
        </button>
      </div>
    </footer>
  </div>
</main>
`;
